pkg/validate/streaming.go

```diff
@@ -0,0 +1,120 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validate
+
+import (
+	"bytes"
+	"io"
+	"net/url"
+
+	"github.com/kubernetes-incubator/cri-tools/pkg/framework"
+	remote "k8s.io/apimachinery/pkg/util/remotecommand"
+	"k8s.io/client-go/rest"
+	"k8s.io/kubernetes/pkg/client/unversioned/remotecommand"
+	internalapi "k8s.io/kubernetes/pkg/kubelet/api"
+	runtimeapi "k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+const (
+	defaultStreamServer string = "http://127.0.0.1:10250"
+)
+
+var _ = framework.KubeDescribe("Streaming", func() {
+	f := framework.NewDefaultCRIFramework()
+
+	var rc internalapi.RuntimeService
+	var ic internalapi.ImageManagerService
+
+	BeforeEach(func() {
+		rc = f.CRIClient.CRIRuntimeClient
+		ic = f.CRIClient.CRIImageClient
+	})
+
+	Context("runtime should support streaming interfaces", func() {
+		var podID string
+		var podConfig *runtimeapi.PodSandboxConfig
+
+		BeforeEach(func() {
+			podID, podConfig = createPodSandboxForContainer(rc)
+		})
+
+		AfterEach(func() {
+			By("stop PodSandbox")
+			rc.StopPodSandbox(podID)
+			By("delete PodSandbox")
+			rc.RemovePodSandbox(podID)
+		})
+
+		It("runtime should support exec [Conformance]", func() {
+			By("test create a default container")
+			containerID := createDefaultContainer(rc, ic, podID, podConfig, "container-for-exec-test")
+
+			By("start container")
+			startContainer(rc, containerID)
+
+			req := createDefaultExec(rc, containerID)
+
+			By("check the output of exec")
+			checkExec(rc, req)
+		})
+	})
+})
+
+func createDefaultExec(c internalapi.RuntimeService, containerID string) string {
+	By("exec default command in container: " + containerID)
+	req := &runtimeapi.ExecRequest{
+		ContainerId: containerID,
+		Cmd:         []string{"echo", "hello"},
+	}
+
+	resp, err := c.Exec(req)
+	framework.ExpectNoError(err, "failed to exec in container %q", containerID)
+	return resp.Url
+}
+
+func checkExec(c internalapi.RuntimeService, req string) {
+	var streamOut, streamErr io.Writer
+	localOut := &bytes.Buffer{}
+	localErr := &bytes.Buffer{}
+
+	streamOut = localOut
+	streamErr = localErr
+
+	url, err := url.Parse(defaultStreamServer + req)
```

Comment:
Should check whether the streaming URL is full URL, it should be a full URL for external runtimes, but dockershim only returns a relative path.